Grant Subscription permission by default

qnkhuat opened the pull request:
Pursuant of #21124, in this PR we want to make sure all groups have Subscription permission enabled by default
To do that, this PR adds a migration script to make sure subscriptions are enabled for all existing permission groups

qnkhuat committed:
add API to fetch general permisisons graph

qnkhuat committed:
add API to update general permissionns

qnkhuat committed:
change author of migration

qnkhuat committed:
merge with master to resolve conflict in migration

qnkhuat committed:
update documents

qnkhuat committed:
misc fixes to applease the CIs

qnkhuat committed:
Add tests for general permission APIs and models

qnkhuat committed:
linting and fix a failed test case

qnkhuat committed:
fix some failed tests

qnkhuat committed:
update docs and change /subscription/ to /general/subscription/ for consistency

qnkhuat committed:
Hook and migration to make sure subscription are created for new groups by default

qnkhuat commented on enterprise/backend/src/metabase_enterprise/advanced_permissions/models/permissions/general_permissions.clj:
```diff
@@ -43,22 +43,9 @@
     (into {} (for [group-id group-ids]
                {group-id (permissions-set-from-group-id group-id)}))))
 
-(s/defn general-perms-path :- perms/Path
```
move this to `src/metabase/models/permissions.clj`

qnkhuat commented on src/metabase/models/permissions_group.clj:
```diff
@@ -100,12 +100,24 @@
+   ((resolve 'metabase.models.permissions/general-perms-path) :subscription)))
+
+(defn- post-insert [group]
+  (u/prog1 group
```
Note that even though General Permission is an enterprise feature, we still grant it for the newly created group even for OSS. 
This makes sure everything will work if users upgrade from OSS to EE

qnkhuat commented on test/metabase/db/schema_migrations_test.clj:
```diff
@@ -491,22 +491,22 @@
             (all-user-perms []
               (db/query {:select [:object], :from [Permissions], :where [:= :group_id (all-users-group-id)]}))]
-      (impl/test-migrations ["v43.00-020"] [migrate!]
+      (impl/test-migrations ["v43.00-020" "v43.00-021"] [migrate!]
```
with the newly added migration, these tests will be affected, thus limiting its end migration.

qnkhuat committed:
add schema migrations tests

qnkhuat committed:
set for the win

camsaul commented on enterprise/backend/test/metabase_enterprise/advanced_permissions/models/permissions/block_permissions_test.clj:
```diff
@@ -113,7 +113,9 @@
               (is (= #{(perms/database-block-perms-path (mt/id))}
-                     (db/select-field :object Permissions :group_id group-id))))))))))
+                     (db/select-field :object Permissions {:where [:and
+                                                                   [:<> :object (perms/general-perms-path :subscription)]
```
Probably better to just check that `object LIKE '/block/%'` or something like that instead so we don't have to keep adding more `object <>` clauses in the future when we add more general perms types

camsaul commented on resources/migrations/000_migrations.yaml:
```diff
@@ -11293,6 +11293,26 @@ databaseChangeLog:
+              SELECT
+                all_users_group.id AS group_id,
+                '/general/subscription/' AS object
+              FROM (
+                SELECT id
+                FROM permissions_group
+                WHERE name <> 'Administrators'
+              ) all_users_group
```
You're selecting any group that's not the admin group, and you're calling it `all_users_group`... that's **super** confusing and misleading. Name it `groups` or `non_admin_groups` or something like that

camsaul commented on src/metabase/models/permissions.clj:
```diff
@@ -451,6 +451,18 @@
   (base->feature-perms-path perm-type perm-value (adhoc-native-query-path db-id)))
 
+(s/defn general-perms-path :- Path
+  "Get general permission's path by permission's type"
```
You should make this a complete sentence https://guide.clojure.style/#docstring-summary

camsaul replied:
Ok, I actually realized you just moved this from the other namespace but it's still worth fixing here

camsaul commented on src/metabase/models/permissions_group.clj:
```diff
@@ -100,12 +100,24 @@
     (when group-name
       (check-name-not-already-taken group-name))))
 
+(defn- grant-subscription-permission
```
```suggestion
(defn- grant-subscription-permission!
```

https://guide.clojure.style/#naming-unsafe-functions

This function has side-effects. This wouldn't be safe in an STM transaction because if you ran it a second time it would fail because that entry would already exist. Group ID + `object` has a unique constraint.

If you're calling functions that end in `!` for side-effects is side of another function, that function's name should end in `!` as well.

camsaul replied:
It also seems to me like **this** function belongs in `metabase.models.permissions` seeing how both `grant-permissions!` and other related similar functions all live there. You'd still have to require-resolve but it would only be one function to `resolve` instead of two

camsaul commented on test/metabase/models/permissions_group_test.clj:
```diff
@@ -104,3 +104,8 @@
+(deftest post-insert-test
+  (testing "Should grant subscription permission for newly created Group"
+    (mt/with-temp PermissionsGroup [{group-id :id}]
+      (is (= 1 (db/count Permissions :group_id group-id :object (perms/general-perms-path :subscription)))))))
```
```suggestion
      (is (db/exists? Permissions :group_id group-id :object (perms/general-perms-path :subscription))))))
```

It's a simpler way of doing the same thing (see the test above for example)

camsaul commented on test/metabase/models/permissions_test.clj:
```diff
@@ -706,7 +706,9 @@
     (letfn [(perms []
-              (db/select-field :object Permissions :group_id group-id))]
+              (db/select-field :object Permissions {:where [:and
+                                                            [:<> :object (perms/general-perms-path :subscription)]
```
Same thing I mentioned about the other test, instead of hardcoding a specific `remove`-style clause like `<> '/general/subscription` it's better to future-proof this and just do a `filter`-style clause that keeps the rows we do want so we don't have to go do this again next time we add a general perm we want to grant everyone by default

camsaul reviewed (dismissed):
A handful of small style notes but overall this looks pretty close

camsaul commented:
It would be good to get @noahmoss 's eyes on this as well since he's probably worked on the permissions code more than anyone at this point

qnkhuat committed:
address noah's comments

qnkhuat committed:
Parse number as is in http-client for tests

qnkhuat committed:
address Cam's comments

qnkhuat committed:
revert the last commit about parsing API response in tests

qnkhuat committed:
Merge branch 'ngoc-general-permission' into ngoc-general-permission-2

qnkhuat committed:
Merge branch 'master' into ngoc-general-permission

qnkhuat committed:
Merge branch 'ngoc-general-permission' into ngoc-general-permission-2

qnkhuat committed:
change fk name

qnkhuat committed:
delete a comment

qnkhuat committed:
Changes:
- Rename `changes` column to `after` to keep things consistent
- If a group  has no General Permisions, it'll not be included in the graph
- Update tests and some docs

qnkhuat committed:
fix failing tests in ee

qnkhuat committed:
merge with ngoc-general-permission

qnkhuat committed:
add some tests and make docstring completes

qnkhuat committed:
Merge branch 'ngoc-general-permission' into ngoc-general-permission-2

qnkhuat committed:
polishing comments

qnkhuat committed:
namespaces

noahmoss commented:
I don't think we actually want to grant subscription permissions for all new groups.

Subscription permissions should be enabled for `All Users` by default. As long as this remains enabled, this will supersede any subscription permissions on individual groups. New groups will _de facto_ have subscription permissions, since all of their members are also members of `All Users`.

If an admin disables subscription permissions for `All Users`, new groups should not have subscription permissions, unless an admin enables them manually.

Does this make sense? Anything I'm missing here? (What I explained above is basically how we already handle data and download perms, if you want to experiment with it)

noahmoss commented on resources/migrations/000_migrations.yaml:
```diff
@@ -11308,6 +11308,26 @@ databaseChangeLog:
+  - changeSet:
+      id: v43.00-047
+      author: qnkhuat
+      comment: Added 0.43.0. Grant permissions to create/edit subscription to all existing groups except Admin
```
Per my top-level comment, I think this migration only needs to add subscription permissions for the `All Users` group.

qnkhuat committed:
address Noah's comments and merge with master

qnkhuat committed:
fix namespaces

qnkhuat committed:
Grant Subscription permission to All Users group only

qnkhuat committed:
Add general permisison flags to `GET /api/user/current` (#21250)

* return general permisison flags for /api/current

* namespaces

* move permission flags to under

qnkhuat committed:
merge with master and fix conflicts

qnkhuat committed:
Enforce Subscription permissions (#21285)

* return general permisison flags for /api/current

* namespaces

* enforce general permissions for subscription and tests

* update check-has-general-permisison to add option to require superuser

* adding arity

qnkhuat committed:
add tests for permissions helper function

noahmoss commented on enterprise/backend/test/metabase_enterprise/advanced_permissions/api/subscription_test.clj:
```diff
@@ -0,0 +1,113 @@
+            [metabase.public-settings.premium-features-test :as premium-features-test]
+            [metabase.test :as mt]))
+
+(defmacro ^:private temporarily-disable-subscription-for-all-users
```
nit: it's not a big deal, but testing macros often have the `with...` naming format. you could call this something like `with-subscription-disabled-for-all-users` for consistency

noahmoss commented on enterprise/backend/test/metabase_enterprise/advanced_permissions/api/subscription_test.clj:
```diff
@@ -0,0 +1,113 @@
+                              (mt/user-http-request user :get status "pulse/form_input"))]
+           (testing "user's group has no subscription permissions"
+             (perms/revoke-general-permissions! group :subscription)
+             (testing "should success if `advanced-permissions` is disabled"
```
nit (below as well)
```suggestion
             (testing "should succeed if `advanced-permissions` is disabled"
```

noahmoss commented on enterprise/backend/test/metabase_enterprise/advanced_permissions/api/subscription_test.clj:
```diff
@@ -0,0 +1,113 @@
+                 (create-pulse 200)
+                 (update-pulse 200)
+                 (get-form 200)))
+             (testing "should failed if `advanced-permissions` is enabled"
```
nit (below as well)
```suggestion
             (testing "should fail if `advanced-permissions` is enabled"
```

noahmoss commented on enterprise/backend/test/metabase_enterprise/advanced_permissions/models/permissions/block_permissions_test.clj:
```diff
@@ -106,7 +106,9 @@
                      (test-db-perms group-id)))
               (is (= #{(perms/database-block-perms-path (mt/id))}
-                     (db/select-field :object Permissions :group_id group-id))))))))))
+                     (db/select-field :object Permissions {:where [:and
+                                                                   [:like :object "/block/%"]
+                                                                   [:= :group_id group-id]]}))))))))))
```
Hmm, I think this change is breaking the purpose of this test?

The test is trying to assert that there _aren't_ data perms, when a group has block perms. But that's not being checked if you're only fetching the block perm path.

You'd have to also include paths like `/db/%`, I think

noahmoss commented on resources/migrations/000_migrations.yaml:
```diff
@@ -1,4 +1,4 @@
-databaseChangeLog:
```
uh, was this intended? I'm pretty sure we don't want indentation here

noahmoss commented on src/metabase/api/common.clj:
```diff
@@ -93,6 +95,21 @@
+  ([perm-type require-superuser?]
+   (classloader/require 'metabase.public-settings.premium-features
+                        'metabase.models.permissions)
+   (if ((resolve 'metabase.public-settings.premium-features/enable-advanced-permissions?))
+     (check-403 (or *is-superuser?*
+                    ((resolve 'metabase.models.permissions/set-has-general-permission-of-type)
+                     @*current-user-permissions-set* perm-type)))
```
I think it would be slightly preferable to have a bit more of the permission checking logic in an EE namespace

ie:
1. Check if the EE ns can be resolved
2. If yes, call into an EE function to do the full permissions checking
3. Else, just check superuser if necessary

noahmoss commented on src/metabase/api/common.clj:
```diff
@@ -93,6 +95,21 @@
+  ([perm-type require-superuser?]
+   (classloader/require 'metabase.public-settings.premium-features
+                        'metabase.models.permissions)
+   (if ((resolve 'metabase.public-settings.premium-features/enable-advanced-permissions?))
+     (check-403 (or *is-superuser?*
+                    ((resolve 'metabase.models.permissions/set-has-general-permission-of-type)
```
neither of these are EE namespaces, why do they need to be resolved?

qnkhuat replied:
yes. it will fail due to cyclic dependencies...

noahmoss replied:
oh I see. maybe you can move this function to `metabase.api.common.validation`? that's a namespace I created recently when I ran into a similar cyclic dependency issue with `metabase.api.common`

qnkhuat replied:
that's fanstatic :)

noahmoss commented on src/metabase/api/user.clj:
```diff
@@ -151,11 +153,24 @@
      :offset offset-paging/*offset*}))
 
 
+(defn- with-general-permissions
+  "Adds to `user` a set of boolean flag indiciate whether or not current user has access to a general permissions.
+  These flags will only be added if `advanced-permissions` is enabled."
+  [user]
+  (if-not (premium-features/enable-advanced-permissions?)
+    user
+    (let [permissions-set @api/*current-user-permissions-set*]
+      (assoc user :permissions
+             {:can_access_setting      (perms/set-has-general-permission-of-type? permissions-set :setting)
+              :can_access_subscription (perms/set-has-general-permission-of-type? permissions-set :subscription)
+              :can_access_monitoring   (perms/set-has-general-permission-of-type? permissions-set :monitoring)}))))
```
I think this logic can also be in an EE namespace. The function here can just be a wrapper around it, if the EE namespace can be resolved. And the `advanced-permissions` feature flag check should live in EE.

noahmoss reviewed (approved):
Looks good. My main comments are about moving more of the permission checking code into EE namespaces, and fixing a block perms test that I think is now not working properly.

Stamping to unblock once these are addressed

Also, I believe merging in the latest master will fix the CI failures

qnkhuat committed:
Move advanced permissions check funcs to ee namespace

qnkhuat committed:
unpushed changes

qnkhuat committed:
namespaces

qnkhuat committed:
ignore exception when load namespaces

qnkhuat committed:
change helper fn name

qnkhuat committed:
Merge branch 'master' into ngoc-general-permission-2

qnkhuat committed:
Merge branch 'master' into ngoc-general-permission-2

qnkhuat committed:
Enforce Monitoring Permissions (#21321)

* return general permisison flags for /api/current

* namespaces

* enforce general permissions for subscription and tests

* update check-has-general-permisison to add option to require superuser

* adding arity

* enforce permissions to call /api/dataset for internal queries

* enforce monitoring permissions for api/task and api/util

* add tests for OSS

* add tests for db-connection-info endpoint

* change test schema

* update name func and fix ns

* whydon't CI run ?

qnkhuat committed:
Enforce Setting Permissions (#21386)

* return general permisison flags for /api/current

* namespaces

* enforce general permissions for subscription and tests

* update check-has-general-permisison to add option to require superuser

* adding arity

* enforce permissions to call /api/dataset for internal queries

* enforce monitoring permissions for api/task and api/util

* add tests for OSS

* add tests for db-connection-info endpoint

* change test schema

* update name func and fix ns

* whydon't CI run ?

* Enforce Setting permissions

* fix failing test

* make sure we could run slack test twice

* make the mock consistent

* address Noah's comments

* shorter permissions check
